feat: implement u256 agglayer asset scaling up procedure

partylikeits1983 opened the pull request:
# Asset Conversion Procedure Overview
This PR adds the functionality for converting between asset amounts with various decimals for Agglayer bridging operations.

Note:
Resolves:  https://github.com/0xMiden/miden-base/issues/2011

The `convert_to_u256_scaled` procedure converts Miden amounts to Agglayer compatible `u256` format for bridging to agglayer chains.

partylikeits1983 commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,89 @@
+
+    drop
+    # => [result]
+end
```
Added a check to assert that `scale <= 18`

partylikeits1983 commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,107 @@
+    # => [AMOUNT_U256[1], AMOUNT_U256[0], SCALE_U256[1], SCALE_U256[0]]
+
+    # Perform U256 multiplication: amount * scale
+    exec.u256::mul_unsafe
```
Unfortunately `u256:: wrapping_mul` isn't available yet in miden-base since its on miden-vm version `v0.20.0`. Will add the comment

partylikeits1983 commented on crates/miden-lib/src/agglayer/utils.rs:
```diff
@@ -0,0 +1,22 @@
+/// Convert Felt values to u32 values.
+pub fn felts_to_u32_slice(felts: &[Felt]) -> Vec<u32> {
+    felts.iter().map(|f| f.as_int() as u32).collect()
+}
```
was a left over, forgot to remove

partylikeits1983 commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,106 @@
+#! - ETH: amount=1e10, target_scale=8 → 1e18
+#!
+#! Invocation: exec
+pub proc convert_felt_to_u256_scaled
```
Maybe renaming to `convert_felt_amount_to_u256_scaled` is even better?

mmagician replied:
I would side with copilot on this one, I don't think it's necessary to mention the data type in the name

partylikeits1983 commented on Cargo.toml:
```diff
@@ -63,9 +63,10 @@ miden-utils-sync      = { default-features = false, version = "0.19" }
-tokio          = { default-features = false, features = ["sync"], version = "1" }
+anyhow          = { default-features = false, features = ["backtrace", "std"], version = "1.0" }
+assert_matches  = { default-features = false, version = "1.5" }
+primitive-types = { default-features = false, version = "0.14" }
```
Importing `U256` type.

mmagician commented on crates/miden-testing/tests/agglayer/asset_conversion.rs:
```diff
@@ -0,0 +1,246 @@
+
+#[test]
+fn test_felts_to_u256_bytes_hand_encoded_values() {
+    // Test case 1: Simple sequential values 1,2,3,4,5,6,7,8
```
I think the test name doesn't need to reflect the fact that these are hand picked

bobbinth commented on crates/miden-lib/src/agglayer/utils.rs:
```diff
@@ -3,6 +3,19 @@ use alloc::vec::Vec;
 use miden_objects::Felt;
 
+/// Convert 8 Felt values (u32 limbs in big-endian order) to U256 bytes in little-endian format.
+pub fn felts_to_u256_bytes(limbs: Vec<Felt>) -> Vec<u8> {
```
I would change the return type to `[u8; 32]` since we know that we will always be returning exactly 32 bytes.

Also, the input could be `&[Felt]` (AFAICT, we don't really care if we get a vector), or better yet, `[Felt; 8]` as we rely on the fact that the input has exactly 8 elements (and will panic otherwise).

Also, would be great to explain in the docs why we do the big-endian to little-endian conversion.

partylikeits1983 replied:
Good idea! Updated to this: 
```rs
pub fn felts_to_u256_bytes(limbs: Vec<Felt>) -> Vec<u8> {
```

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,106 @@
+#! both the amount and scale factor to U256 format, performs U256 multiplication, and
+#! returns the result as 8 u32 limbs ready for Agglayer bridging operations.
+#!
+#! Inputs:  [amount, target_scale]
```
I think it is fine as is.

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,106 @@
+#! Where:
+#! - amount: The asset amount to be converted
+#! - target_scale: Exponent for scaling factor (10^target_scale)
+#! - [RESULT_U256[1], RESULT_U256[0]]: U256 value as 8 u32 limbs in big-endian order
```
Could we clarify what big-ending here means to make it un-ambiguous? For example, does it mean that the most significant limb is at the top of the stack?

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,106 @@
+#! - ETH: amount=1e10, target_scale=8 → 1e18
+#!
+#! Invocation: exec
+pub proc convert_amount_to_u256_scaled
```
nit: I would maybe rename this to something like `scale_native_amount_to_u256`.

partylikeits1983 replied:
Renamed to `scale_native_amount_to_u256`

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,106 @@
+    # Perform U256 multiplication: amount * scale
+    # This is safe because both the scaling factor and amount are guaranteed to be smaller
+    # than 2^64, so we will never overflow a 256-bit value.
+    exec.u256::mul_unsafe
```
Thinking about this more: since both scale and amount are u64 values, could we not just use u64 multiplication (the overflowing variant)? For example, something like:

```
# => [amount_hi, amount_lo, scale_hi, scale_lo]

exec.u64::overflowing_mul
# => [res_hi, res_mid_hi, res_mid_lo, res_lo]

push.0.0.0.0
# => [RESULT_U256[1], RESULT_U256[0]]
```

This should be much more efficient than using `u256` multiplication.

bobbinth reviewed (commented):
Looks good! Thank you! I left a few comments inline - the main one is about using `u64` multiplication to make the procedure much more efficient.

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,105 @@
+    # => [res_hi, res_mid_hi, res_mid_lo, res_lo]
+
+    push.0.0.0.0
+    # => [RESULT_U256[1], RESULT_U256[0]]
```
Thinking a bit more about this now, I wonder if we should change the order to output the result in little-endian format. It seems like we need to do this anyway, and overall, we are moving in that direction with some of the recent and upcoming changes in the VM. We could do it like so:

```
exec.u64::overflowing_mul
exec.word::reverse
padw swapw
# => [RESULT_U256[0], RESULT_U256[1]]
```

bobbinth commented on crates/miden-lib/asm/agglayer/account_components/asset_conversion.masm:
```diff
@@ -0,0 +1,105 @@
+
+    # truncate stack to 8 felts representing U256
+    movupw.2 dropw movup.8 drop movup.8 drop
+    # => [RESULT_U256[1], RESULT_U256[0]]
```
For procedures that are called using the `exec` instruction, we should not be truncating the stack - this should be the responsibility of the caller. Otherwise, we may drop things from the stack that the caller does not expect to be dropped.

So, let's remove the truncation code from here.

partylikeits1983 commented on crates/miden-lib/src/agglayer/utils.rs:
```diff
@@ -3,6 +3,23 @@ use alloc::vec::Vec;
+/// - Ethereum/EVM expects U256 values as 32 bytes in little-endian format (least significant byte
+///   first)
+/// - This ensures compatibility when bridging assets between Miden and Ethereum-based chains
+pub fn felts_to_u256_bytes(limbs: [Felt; 8]) -> Vec<u8> {
```
Updated to return little-endian order

partylikeits1983 commented:
Changed `felts_to_u256_bytes` return type from `Vec<u8>` to `[u8; 32]` and updated it to accept little-endian input order. Updated asset_conversion.masm to output U256 results in little-endian format and fixed stack comments to properly document intermediate values. Updated corresponding tests to match the new little-endian behavior across both Rust and MASM.

I think this is ready to merge @bobbinth @mmagician

bobbinth reviewed (approved):
All looks good! Thank you!
